### ISSUE TITLE
ReduceLROnPlateau does not recognise val_loss despite progress_bar dict

### ISSUE THREAD

**Issue by LucFrachon** [PyTorchLightning/pytorch-lightning#1156]
## 🐛 Bug
When training my model, I get the following message:
```
  File "C:\Users\Luc\Miniconda3\envs\pytorch\lib\site-packages\pytorch_lightning\trainer\training_loop.py", line 371, in train
    raise MisconfigurationException(m)
pytorch_lightning.utilities.debugging.MisconfigurationException: ReduceLROnPlateau conditioned on metric val_loss which is not available. Available metrics are: loss
```
Ihis is similar to #321for instance, but I definitely return a `progress_bar` dict with a `val_loss` key in it (see code below).

#### Code sample

 ```
   def training_step(self, batch, batch_idx):
        z, y_true = batch
        y_pred = self.forward(z)
        loss_val = self.loss_function(y_pred, y_true)
        return {'loss': loss_val.sqrt()}

    def validation_step(self, batch, batch_idx):
        z, y_true = batch
        lr = torch.tensor(self.optim.param_groups[0]['lr'])
        y_pred = self.forward(z)
        loss_val = self.loss_function(y_pred, y_true)
        return {'val_loss': loss_val.sqrt(), 'lr': lr}

    def validation_epoch_end(self, outputs):
        val_loss_mean = torch.stack([x['val_loss'] for x in outputs]).mean()
        lr = outputs[-1]['lr']
        logs = {'val_loss': val_loss_mean, 'lr': lr}
        return {'val_loss': val_loss_mean, 'progress_bar': logs, 'log': logs}
```

### Expected behavior
The `val_loss` value should be picked up by the progress bar.

### Environment

 - PyTorch Version (e.g., 1.0): 1.4.0
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6.10
 - CUDA/cuDNN version: 10
 - GPU models and configuration: 1070Ti x 1
 - Any other relevant information:

**Comment by SkafteNicki** [comment]
Actually, `ReduceLROnPlateau` lr schedulers are conditioned on `callback_metrics`, see here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/774d9be3577ce1f05d895134d5ce70c16877d08b/pytorch_lightning/trainer/training_loop.py#L717-L726
and these values are everything not defined in `'progress_bar'`, `'log'` and `'hidden'`, see here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/774d9be3577ce1f05d895134d5ce70c16877d08b/pytorch_lightning/trainer/logging.py#L106-L109
since you are returning `'val_loss'` as a separate key this should however not be the problem.

Since only `'loss'` is available as a `callback_metrics` it seems that your lr scheduler is called before your validation data is processed. What does your `configure_optimizers()` look like?

**Comment by LucFrachon** [comment]
Thanks for your answer. If I understand correctly, `ReduceLROnPlateau` is the only LR scheduler that does not use values from `progress_bar`?

Here's `configure_optimizers()`:

```
    def configure_optimizers(self):
        self.optim = torch.optim.AdamW(self.parameters(), lr=self.hp.lr_ini, eps=1.e-4)
        self.scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(self.optim, 'max', factor=self.hp.lr_reduce_factor,
                                                                    patience=self.hp.lr_reduce_patience,
                                                                    min_lr=self.hp.lr_min)
        return [self.optim], [self.scheduler]
```

And here is the `Trainer` definition:

`trainer = pl.Trainer(gpus=-1, early_stop_callback=None, max_epochs=stgs.PRED_HPARAMS['max_epochs'])`

**Comment by SkafteNicki** [comment]
Only `ReduceLROnPlateau` schedulers are conditioned on some value. All other LR schedulers have nothing to do with `callback_metrics` or `progress_bar`. Hope this answers your question.

I don't see any obvious mistakes in your code. As a said in my earlier comment, my best guess is that your `ReduceLROnlr scheduler is called before `val_loss` is actually calculated. You probably need to check this.

**Comment by alexeykarnachev** [comment]
I have the same issue. That's how I deal with it:

```python
    def training_step(self, batch, batch_idx):
        loss, lm_loss, mc_loss = self.forward(batch)
        lr = self.trainer.optimizers[0].param_groups[0]['lr']
        lr = torch.tensor(lr).unsqueeze(0).to(loss.device) if len(loss.size()) else lr
        log = {
            'MC-Loss/train': mc_loss,
            'LM-Loss/train': lm_loss,
            'Learning-Rate': lr
        }
        # Set up placeholders for valid metrics.
        if self.trainer.global_step == 0:
            log.update({'LM-Loss/valid': np.inf})

        return {'loss': loss, 'log': log}
```

here I check `if self.trainer.global_step == 0` and if yes, I set up a placeholder.

@SkafteNicki maybe you can advice a better way

**Comment by LucFrachon** [comment]
> Only `ReduceLROnPlateau` schedulers are conditioned on some value. All other LR schedulers have nothing to do with `callback_metrics` or `progress_bar`. Hope this answers your question.
> 
> I don't see any obvious mistakes in your code. As a said in my earlier comment, my best guess is that your `ReduceLROnlr scheduler is called before `val_loss` is actually calculated. You probably need to check this.

@SkafteNicki Not sure I follow you. Maybe I misunderstood, but doesn't PL automatically handle calls to the scheduler? Doesn't it defeat (some of) the purpose of PL if I have to call it myself?

@alexeykarnachev Thanks, I'll try that, but maybe there is a less hacky way...?

**Comment by SkafteNicki** [comment]
@LucFrachon sorry, I must have misinterpret what you asked about. Yes, PL automatically calls the scheduler you define. Sorry for the confusion.

@alexeykarnachev I think that indicates that there is some bug here, since you need to setup a placeholder for the first step. I will see if I can come up with a more permanent solution.

**Comment by alexeykarnachev** [comment]
@SkafteNicki , btw, the trick:
```python
        if self.trainer.global_step == 0:
            log.update({'Loss/valid': np.inf})
```
doesn't help in case of warm start from checkpoint, because in warm start, the `global_step` is newer equal to 0 :)

**Comment by SkafteNicki** [comment]
@alexeykarnachev I have a hard time reproducing this bug, it seems to be a very corner case. Do you have a simple model that I can reproduce?

**Comment by alexeykarnachev** [comment]
@SkafteNicki 
Yes, sure:
https://gist.github.com/alexeykarnachev/efbd40f0ff0cfcd1e324de044a802c25

**Comment by SkafteNicki** [comment]
Okay, after looking at your code @alexeykarnachev, this does not seems to be a bug. When you set `interval': 'step'` you are calling the `.step()` method for `ReduceLROnPlateau` after each batch and it therefore makes complete sense that no `val_loss` is calculated yet. If you really want to do something like this, you need to set `val_check_interval` in the `Trainer` construction to a number lower than `frequency` in the scheduler construction. In this way `val_loss` will be calculated before `.step()` is called.

**Comment by alexeykarnachev** [comment]
Thx, now it's more clear.
Does it mean, that I can't use a schedulers, which are depend on some valid metric with `interval="step"`?
For instance, I have a custom reduce on plateau scheduler, which just ignores nan metric values, that's why I need to set a placeholder for valid metric (like in my first example).

**Comment by SkafteNicki** [comment]
I do not think it is possible just out of the box. However, if you configure your scheduler correctly, then it should be possible. For example, if I initialize my Trainer as
`trainer = Trainer(val_check_interval=50)`
and initialize my scheduler as
```
scheduler = {
    'schduler': ReduceLROnPlateau(optimizer, mode, factor, patience),
    'interval': 'step',
    'frequency': 100
}
```
it should work (not tested), since `val_loss` will be created every 50 steps but the scheduler will first be called after 100 steps.

**Comment by alexeykarnachev** [comment]
oh, right, I missed the `frequency` argument. Totally forgot about it. I'll try it. Thank you!

**Comment by Borda** [comment]
shall be fixed with #1989
